apps/anoma_lib/lib/examples/enock.ex

```diff
@@ -2765,4 +2765,155 @@ defmodule Examples.ENock do
+  ####################################################################
+
+  # Write Program Generator (writes a constant value to a key)
+  def write_code_gen(key_int, value) do
+    reservations = [1 | key_int]
+    writes_logic = [1 | [key_int | value]]
+    [[0 | 3] | [reservations | [writes_logic | [0 | 0]]]]
+  end
+
+  @doc """
+  Nock program for reading "a", "b", adding, and writing sum to "c". Used in EShardBackend tests.
+  """
+  @spec read_ab_write_c_sum() :: Noun.t()
+  def read_ab_write_c_sum() do
+    key_a = a_int()
+    key_b = b_int()
+    key_c = c_int()
+
+    # [[0 | a], [0 | b] | [1 | c]]
+    reservations = abc_list()
+
+    scryA = [12 | [[1 | 0] | [1 | key_a]]]
+    scryB = [12 | [[1 | 0] | [1 | key_b]]]
+    addFormula = [4 | [4 | [4 | [4 | [0 | 2]]]]]
+    sumFormula = [7 | [[scryA | scryB] | addFormula]]
+    writes_logic = [[1 | key_c] | sumFormula]
+
+    [[0 | 3] | [reservations | [writes_logic | [0 | 0]]]]
+  end
+
+  @doc """
+  Nock program for copying value from "a" to "b". Used in EShardBackend tests.
+  """
+  @spec copy_a_to_b() :: Noun.t()
+  def copy_a_to_b() do
+    key_a = a_int()
+    key_b = b_int()
+
+    reservations = [[0 | key_a] | [1 | key_b]]
+    scryA = [12 | [[1 | 0] | [1 | key_a]]]
+    writes_logic = [[1 | key_b] | scryA]
+
+    [[0 | 3] | [reservations | [writes_logic | [0 | 0]]]]
+  end
+
+  @doc """
+  Nock program that reserves write on 'a', then crashes. Used in EShardBackend tests.
+  """
+  @spec crash_after_reserve_a() :: Noun.t()
+  def crash_after_reserve_a() do
+    reservations = [1 | a_int()]
+    # Invalid Nock [0 | [0 | 0]] for the writes stage causes a crash
+    [[0 | 3] | [reservations | [0 | [0 | 0]]]]
+  end
+
+  @doc """
+  I test a Nock program for writing 3 to "a".
+
+  I verify stage 1 (reservation/code split) and stage 2 (execution).
+  """
+  @spec test_nock_program_tx1() :: :ok
+  def test_nock_program_tx1() do
+    # Setup
+    dummy_tx_id = "test_tx_id_1"
+    tx_code = write_code_gen(a_int(), 3)
+    expected_reservations = [1 | a_int()]
+    expected_writes = [a_int() | 3]
+
+    # Stage 1: Execute the core using the backend's formula to extract reservations
+    {:ok, stage1_result} = Nock.nock(tx_code, [9, 2, 0 | 1], %Nock{})
+
+    # Extract components from stage 1 result
+    [reservations | stage2_code] = stage1_result
+    assert reservations == expected_reservations
+
+    # Stage 2: Execute using the exact sequence from Backends.vm_execute_stage2
+    env_tx1 = %Nock{}
+
+    # Step 1: Apply formula [10, [6, 1 | id], 0 | 1] - this inserts the tx_id into the code
+    {:ok, ordered_tx1} =
+      Nock.nock(stage2_code, [10, [6, 1 | dummy_tx_id], 0 | 1], env_tx1)
+
+    # Step 2: Apply formula [9, 2, 0 | 1] - this executes the transaction
+    {:ok, result1} = Nock.nock(ordered_tx1, [9, 2, 0 | 1], env_tx1)
+
+    # This is the expected format that backends.ex receives from vm_execute_stage2
+    assert result1 == expected_writes
+
+    :ok
+  end
```

Comment:
Shouldn't we return the `%Nock{}` datastructure.

In fact you can use this setup for further examples that build upon this